### PR TITLE
Use caching walker in procspy

### PR DESCRIPTION
**Pull request by tomwilkie**
Builds on #737, <s>will rebase once thats in</s>.  Part of #284.

Gives us a 1-2% cpu usage reduction on my VM.

### CODE REVIEW

File: common/fs/fs.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	"io"
+	"io/ioutil"
+	"os"
+	"syscall"
+)
+
+// T is the filesystem interface type.
+type T interface {
```
Comment (paulbellamy) [inline common/fs/fs.go]: In the `sort` package (and other places?) this is called `type Interface interface {...}`. Not sure if I like that, but it's a bit more descriptive than `T`.

File: common/fs/fs.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	ReadFile(string) ([]byte, error)
+	Lstat(string, *syscall.Stat_t) error
+	Stat(string, *syscall.Stat_t) error
+	Open(string) (io.ReadWriteCloser, error)
+}
+
+type realFS struct{}
+
+// FS is the way you should access the filesystem.
+var FS T = realFS{}
```
Comment (paulbellamy) [inline common/fs/fs.go]: Maybe move this (and `type realFS ...`) below the mocking stuff so it's all together.

File: probe/process/walker_linux.go [hunk 3]
```diff
@@ -28,7 +23,7 @@ func NewWalker(procRoot string) Walker {
 // passes one-by-one to the supplied function. Walk is only made public
 // so that is can be tested.
 func (w *walker) Walk(f func(Process)) error {
-	dirEntries, err := ReadDir(w.procRoot)
+	dirEntries, err := fs.FS.ReadDir(w.procRoot)
```
Comment (paulbellamy) [inline probe/process/walker_linux.go]: Could have:
```Go
func ReadDir(...) {
  return FS.ReadDir
}
```
And that would clean up the invocations.

**Comment by paulbellamy** [comment]
Other than the sugar/naming quibbles, it LGTM.

File: common/fs/fs.go (hunk 1 above)
Comment (tomwilkie) [inline common/fs/fs.go]: Good idea.

File: common/fs/fs.go (hunk 2 above)
Comment (tomwilkie) [inline common/fs/fs.go]: Roger.

File: probe/process/walker_linux.go (hunk 3 above)
Comment (tomwilkie) [inline probe/process/walker_linux.go]: Yes thats also a good idea - on a roll!

**Pull request merged**